lib/lzo.ml

```diff
@@ -0,0 +1,247 @@
+  | '\000', Zero ->
+    junk byte >>= fun () -> count >>= fun count ->
+    let len = 3 + 15 + count in
+    transmit ~len None >>= fun () -> m
+  | '\000' .. '\015', ( One | Two | Three ) ->
+    let d, state =
+      Char.code chr lsr 2,
+      state_of_int (Char.code chr land 0b11) in
+    junk byte >>= fun () -> peek byte >>= fun h ->
+    junk byte >>= fun () ->
+    let off = (Char.code h lsl 2) + d + 1 in
+    copy ~off ~len:2 state >>= fun () -> m
+  | '\000' .. '\015', None ->
+    peek byte >>= fun h ->
+    junk byte >>= fun () ->
+    let state = state_of_int (Char.code chr land 0b11) in
+    let off = (Char.code h lsl 2) + (Char.code chr lsr 2) + 2049 in
+    copy ~off ~len:3 state >>= fun () -> m
+  | '\016' .. '\031', _ ->
+    let length = Char.code chr land 0b111 in
+    junk byte >>= fun () ->
+    let with_length len =
+      peek leshort >>= fun s ->
+      junk leshort >>= fun () ->
+      let state = state_of_int (s land 0b11) in
+      let off =
+        let h = (Char.code chr land 8) lsr 3 in
+        16384 + (h lsl 14) + (s lsr 2) in
+      copy ~off ~len state >>= fun () ->
+      if off = 16384 then end_of_lzo else m in
+    if length = 0
+    then count >>= fun count -> with_length (2 + 7 + length)
+    else with_length length
+  | '\032' .. '\063', _ ->
+    let with_length len =
+      peek leshort >>= fun s ->
+      let state = state_of_int (s land 0b11) in
+      let off = succ (s lsr 2) in
+      copy ~off ~len state >>= fun () -> m in
+    let length = Char.code chr land 0b11111 in
+    junk byte >>= fun () ->
+    if length = 0
+    then count >>= fun count -> with_length (2 + 31 + count)
+    else with_length (2 + length)
+  | '\064' .. '\255', _ ->
+    let state = state_of_int (Char.code chr land 0b11) in
+    let len, d =
+      let shift = (Char.code chr lsr 7) lsl 1 in
+      3 + shift + ((Char.code chr lsr 5) land (1 lor shift)),
+      (Char.code chr lsr 2) land 0b111 in
+    peek byte >>= fun h ->
+    junk byte >>= fun () ->
+    let off = (Char.code h lsl 3) + d + 1 in
+    copy ~off ~len state >>= fun () -> m
+
+let fiber : (unit, error) t = peek byte >>= function
+  | '\016' -> malformedf "No dictionary at offset 0 available"
+  | '\000' .. '\017' -> fiber
+  | '\018' -> junk byte >>= fun () -> transmit ~len:1 Zero >>= fun () -> fiber
+  | '\019' -> junk byte >>= fun () -> transmit ~len:2 Zero >>= fun () -> fiber
+  | '\020' -> junk byte >>= fun () -> transmit ~len:3 Zero >>= fun () -> fiber
+  | '\021' -> junk byte >>= fun () -> transmit ~len:4 Zero >>= fun () -> fiber
+  | '\022' .. '\255' as chr ->
+    let len = Char.code chr - 17 in
+    junk byte >>= fun () -> transmit ~len None
+
+let uncompress input : (bigstring, error) result =
+  let max = Int64.(mul (of_int (bigstring_length input)) 256L) in
+  if max > Int64.of_int Sys.max_string_length
+  then invalid_arg "Deflated input is too huge" ;
+  let output = bigstring_create (Int64.to_int max) in
```

Comment:
In `openvpn` (where we needed LZO originally) we know the max output size the compress payload is a network packet that has to conform to a known MTU (usually ~1500 bytes), so we could provide that to prevent wasting time / memory on bad inputs.

In scenarios like that, maybe letting the user optionally allocate + provide the `output` buffer would make sense? (Then we can re-use the same output buffer for successive calls, or at least set max size and fail early) - that also avoids allocating 256x input size buffers (which is the *upper* bound for the ideal compression ratio; for most real-world inputs that is probably excessive).